packages/plugin-ext/src/plugin/tree/tree-views.ts

```diff
@@ -346,18 +345,32 @@ class TreeViewExtImpl<T> implements Disposable {
         return idLabel;
     }
 
-    private buildTreeItemId(parentId: string, item: TreeItem): string {
-        // build tree id according to https://code.visualstudio.com/api/references/vscode-api#TreeItem
-        // note: the front end tree implementation cannot handle reparenting items, hence the id is set to the "path" of individual ids
-        let id = typeof item.id === 'string' ? item.id : this.getItemLabel(item);
-        if (id) {
-            // we use '' as the id of the root, we don't consider that a valid id
-            // since '' is falsy, we'll never get '' in this branch
-            id = TreeViewExtImpl.ID_ITEM + id;
-        } else {
-            id = TreeViewExtImpl.ID_COMPUTED + this.nextItemId++;
+    // Modeled on https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/common/extHostTreeViews.ts#L822
+    private buildTreeItemId(parentId: string, item: TreeItem, mustReturnNew: boolean): string {
+        if (item.id) {
+            return `${TreeViewExtImpl.ID_ITEM}/${item.id}`;
         }
-        return `${parentId}/${id}`;
+
+        const treeItemLabel = this.getItemLabel(item);
+        const prefix: string = parentId ? parentId : TreeViewExtImpl.ID_COMPUTED;
+        let elementId = treeItemLabel ? treeItemLabel : item.resourceUri ? basename(item.resourceUri.fsPath) : '';
+        elementId = elementId.indexOf('/') !== -1 ? elementId.replace('/', '//') : elementId;
+        const childrenNodes = (this.nodes.get(parentId)?.children || []);
+
+        let id: string;
+        let counter = 0;
```

Comment:
This will generate the same ids potentially different domain elements, because it's based on the position of an element in the parent. I think this is wrong: we're assuming sameness, when in fact, we don't know. A simple fix would be to have a global counter instead of starting at 0 here.
The API doc states

> If not provided, an id is generated using the tree item's label. Note that when labels change, ids will change and that selection and expansion state cannot be kept stable anymore.

Note the absence of "position" as a as a part of the computed ID.

Reply:
I think the _really_ correct thing to do is to uniqify any two computed id's that end up the same and otherwise treat them as a true id.

Reply:
If `mustReturnNew` is true, then it won't generate the same id's for different items. The current children can't be using all of the counter values up to the current length of the children array; either there's a gap in the middle, or the value at `.length` itself must be open.

Reply:
But what about the next time we fetch the children? We will generate the same id for a potentially different element, just because it's in the same place?

Reply:
Yes, that is possible. As I noted below, the promise of certainty with computed ID's is lower than with plugin-supplied ID's. There's no way around that. Since we depend on something that is not guaranteed to be unique, and indeed is demonstrably not unique in a number of cases - it is impossible to guarantee that we will always match the same item. But matching an item that is approximately the same may be better than matching no item at all. That seems to be the design choice that VSCode has made.

Reply:
That whole "computed id" business is a disaster in the VS Code API, if you ask me, but I buy your argument.